### PR TITLE
Optimize `try_expand_impl_trait_type`

### LINKED ISSUE #65147: check_mod_item_types query times regressed bigly
There is a crate inside of Fuchsia whose compile times increased from ~500ms to ~120s in the last month of Rust compiler updates. It hasn't been easy so far to provide a minimal example, but I will keep working at it.

Some bisecting revealed that it actually consisted of two performance regressions, both affecting the time spent in `check_mod_item_types`:

| Commit hash                              | Item                                     | Self time | % of total time | Item count | Cache hits | Blocked time | Incremental load time |
|------------------------------------------|------------------------------------------|-----------|-----------------|------------|------------|--------------|-----------------------|
| 87b0c9036f0557568cfc939542508e9adf874b83 | check_mod_item_types                     | 34.92ms   | 7.120           | 6          | 0          | 0.00ns       | 0.00ns                |
| 34e82a7b793a6cdd27df762bf46bab8cdc92b14a | check_mod_item_types                     | 21.92s    | 97.786          | 6          | 0          | 0.00ns       | 0.00ns                |
| 66bf391c3aabfc77f5f7139fc9e6944f995d574e | check_mod_item_types                     | 21.76s    | 97.838          | 6          | 0          | 0.00ns       | 0.00ns                |
| 0221e265621a5fcc68ca62bdcdeabad1882a0e9a | check_mod_item_types                     | 133.53s   | 99.526          | 6          | 0          | 0.00ns       | 0.00ns                |

The first regression occurred in the rollup https://github.com/rust-lang/rust/pull/64354, and the second occurred sometime in the two weeks between changes 66bf391c3aabfc77f5f7139fc9e6944f995d574e..0221e265621a5fcc68ca62bdcdeabad1882a0e9a.

There is heavy use of async/await in the code, and I've identified a couple places where taking out `.await` makes the compile finish right away. This makes me suspect https://github.com/rust-lang/rust/pull/64292 for the first regression, but I'm not sure.

I'll update the issue as I have more information.

### PR DESCRIPTION
**Pull request by tmandry**
A lot of time was being spent expanding some large `impl Future` types in fuchsia. This PR takes the number of types being visited in one expansion from >3 billion to about a thousand, and eliminates the compile time regression in https://github.com/rust-lang/rust/issues/65147 (in fact, compile times are better than they were before).

Thanks to @Mark-Simulacrum for helping identify the issue and to @matthewjasper for suggesting this change. 

Fixes #65147.
r? @matthewjasper,@nikomatsakis

### CODE REVIEW

**Comment by tmandry** [comment]
@rust-timer queue

**Comment by tmandry** [comment]
@rust-timer build 0d2ff89f766a2f16a09c41ee7c65453980e81908

File: src/librustc/ty/util.rs [hunk 1]
```diff
@@ -697,6 +697,9 @@ impl<'tcx> TyCtxt<'tcx> {
             // that type, and when we finish expanding that type we remove the
             // its DefId.
             seen_opaque_tys: FxHashSet<DefId>,
+            // Cache of all expansions we've seen so far. This is a critical
+            // optimization for some large types produced by async fn trees.
```
Comment (Centril) [inline src/librustc/ty/util.rs]:
```suggestion
            // optimization for some large types produced by `async fn` trees.
```

File: src/librustc/ty/util.rs [hunk 2]
```diff
@@ -713,10 +716,17 @@ impl<'tcx> TyCtxt<'tcx> {
                 }
                 let substs = substs.fold_with(self);
                 if self.seen_opaque_tys.insert(def_id) {
-                    let generic_ty = self.tcx.type_of(def_id);
-                    let concrete_ty = generic_ty.subst(self.tcx, substs);
-                    let expanded_ty = self.fold_ty(concrete_ty);
-                    self.seen_opaque_tys.remove(&def_id);
+                    let expanded_ty = match self.expanded_cache.get(&def_id) {
```
Comment (Centril) [inline src/librustc/ty/util.rs]: Why not use the `.entry` API?

File: src/test/ui/impl-trait/recursive-impl-trait-type--through-non-recursize.stderr [hunk 3]
```diff
@@ -20,15 +20,15 @@ error[E0720]: opaque type expands to a recursive type
 LL | fn recursive_wrap() -> impl Sized {
    |                        ^^^^^^^^^^ expands to a recursive type
    |
-   = note: expanded type is `((impl Sized,),)`
+   = note: expanded type is `(impl Sized,)`
```
Comment (Centril) [inline src/test/ui/impl-trait/recursive-impl-trait-type--through-non-recursize.stderr]: What's going on here?

**Comment by nikomatsakis** [comment]
@tmandry I'm not sure if we have a suitable test already, but if not, can you add something to `perf` that reflects this compile time gap?

File: src/librustc/ty/util.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/librustc/ty/util.rs]: Yeah! Entry! Entry! I ❤️ Entry!

Comment (nikomatsakis) [reply src/librustc/ty/util.rs]: Oh, I think I see why not -- the `self.fold_ty` call would error out with `entry`, wouldn't it?

**Review by nikomatsakis — APPROVED** [review]
Presuming this helps, and if we add a perf regression test, seems good to me! Nice detective work y'all.

**Comment by tmandry** [comment]
> @tmandry I'm not sure if we have a suitable test already, but if not, can you add something to `perf` that reflects this compile time gap?

Will do. I was hoping something would turn up in the existing test suite, but it doesn't look that way :)

File: src/librustc/ty/util.rs [hunk 4]
```diff
@@ -697,6 +697,9 @@ impl<'tcx> TyCtxt<'tcx> {
             // that type, and when we finish expanding that type we remove the
             // its DefId.
             seen_opaque_tys: FxHashSet<DefId>,
+            // Cache of all expansions we've seen so far. This is a critical
+            // optimization for some large types produced by async fn trees.
+            expanded_cache: FxHashMap<DefId, Ty<'tcx>>,
```
Comment (matthewjasper) [inline src/librustc/ty/util.rs]: The `SubstRef` also should be part of the key here.

File: src/librustc/ty/util.rs [hunk 5]
```diff
@@ -713,10 +716,17 @@ impl<'tcx> TyCtxt<'tcx> {
...
-                    let concrete_ty = generic_ty.subst(self.tcx, substs);
-                    let expanded_ty = self.fold_ty(concrete_ty);
-                    self.seen_opaque_tys.remove(&def_id);
+                    let expanded_ty = match self.expanded_cache.get(&def_id) {
+                        Some(expanded_ty) => expanded_ty,
+                        None => {
+                            let generic_ty = self.tcx.type_of(def_id);
+                            let concrete_ty = generic_ty.subst(self.tcx, substs);
+                            let expanded_ty = self.fold_ty(concrete_ty);
+                            self.seen_opaque_tys.remove(&def_id);
```
Comment (matthewjasper) [inline src/librustc/ty/util.rs]: This should be in both arms

**Review by matthewjasper — CHANGES_REQUESTED** [review]
I think the following test case shows the problem here:

```rust
#![type_length_limit="10000000"]
#![allow(unused)]

macro_rules! mk_async_fn {
    ($f:ident $g:ident) => {
        async fn $g() -> i32 {
            $f().await;
            $f().await;
            $f().await
        }
    }
}

async fn a() -> i32 { 1 }

mk_async_fn!(a b);
mk_async_fn!(b c);
mk_async_fn!(c d);
mk_async_fn!(d e);
mk_async_fn!(e f);
mk_async_fn!(f g);
mk_async_fn!(g h);
mk_async_fn!(h i);
mk_async_fn!(i j);
```

File: src/librustc/ty/util.rs (hunk 2 above)
Comment (tmandry) [reply src/librustc/ty/util.rs]: Yeah, the fact that the closure would need `&mut self` prevents me from using `entry` here :/

File: src/test/ui/impl-trait/recursive-impl-trait-type--through-non-recursize.stderr (hunk 3 above)
Comment (tmandry) [reply src/test/ui/impl-trait/recursive-impl-trait-type--through-non-recursize.stderr]: This no longer changes.

**Comment by tmandry** [comment]
@matthewjasper thanks for supplying the benchmark!

This should be ready, pending another timer run once https://github.com/rust-lang-nursery/rustc-perf/pull/508 is merged.

**Comment by Mark-Simulacrum** [comment]
This timing run should include a diff of the newly-added benchmark.

**Review by matthewjasper — APPROVED** [review]

**Pull request merged**